File: apps/server/app/schemas/entities.py
from app.schemas.base import Contract
from app.schemas.common import AssetClass, TransactionType


class Holding(Contract):
    id: str
    portfolio_id: str
    symbol: str
    name: str
    asset_class: AssetClass
    sector: str
    quantity: float
    cost_basis: float
    current_price: float
    seed: bool = False


class Transaction(Contract):
    id: str
    portfolio_id: str
    symbol: str
    type: TransactionType
    quantity: float
    price: float
    timestamp: str
    seed: bool = False


class Portfolio(Contract):
    id: str
    name: str
    base_currency: str
    created_at: str
    seed: bool = False


class AllocationSlice(Contract):
    key: str
    value: float
    weight: float


class PositionMetric(Contract):
    symbol: str
    name: str
    value: float
    weight: float
    cost_basis: float
    unrealized_gain: float
    unrealized_gain_pct: float


class PortfolioMetrics(Contract):
    portfolio_id: str
    total_value: float
    total_cost_basis: float
    unrealized_gain: float
    unrealized_gain_pct: float
    positions: list[PositionMetric]
    sector_allocation: list[AllocationSlice]
    asset_class_allocation: list[AllocationSlice]
    top_positions: list[PositionMetric]


class PortfolioDetail(Contract):
    portfolio: Portfolio
    holdings: list[Holding]
    metrics: PortfolioMetrics

File: packages/types/src/entities.ts
import { z } from "zod";
import { assetClassSchema, transactionTypeSchema } from "./common.js";

export const holdingSchema = z.object({
  id: z.string(),
  portfolioId: z.string(),
  symbol: z.string(),
  name: z.string(),
  assetClass: assetClassSchema,
  sector: z.string(),
  quantity: z.number(),
  costBasis: z.number(),
  currentPrice: z.number(),
  seed: z.boolean().default(false),
});
export type Holding = z.infer<typeof holdingSchema>;

export const transactionSchema = z.object({
  id: z.string(),
  portfolioId: z.string(),
  symbol: z.string(),
  type: transactionTypeSchema,
  quantity: z.number(),
  price: z.number(),
  timestamp: z.string(),
  seed: z.boolean().default(false),
});
export type Transaction = z.infer<typeof transactionSchema>;

export const portfolioSchema = z.object({
  id: z.string(),
  name: z.string(),
  baseCurrency: z.string(),
  createdAt: z.string(),
  seed: z.boolean().default(false),
});
export type Portfolio = z.infer<typeof portfolioSchema>;

export const allocationSliceSchema = z.object({
  key: z.string(),
  value: z.number(),
  weight: z.number(),
});
export type AllocationSlice = z.infer<typeof allocationSliceSchema>;

export const positionMetricSchema = z.object({
  symbol: z.string(),
  name: z.string(),
  value: z.number(),
  weight: z.number(),
  costBasis: z.number(),
  unrealizedGain: z.number(),
  unrealizedGainPct: z.number(),
});
export type PositionMetric = z.infer<typeof positionMetricSchema>;

export const portfolioMetricsSchema = z.object({
  portfolioId: z.string(),
  totalValue: z.number(),
  totalCostBasis: z.number(),
  unrealizedGain: z.number(),
  unrealizedGainPct: z.number(),
  positions: z.array(positionMetricSchema),
  sectorAllocation: z.array(allocationSliceSchema),
  assetClassAllocation: z.array(allocationSliceSchema),
  topPositions: z.array(positionMetricSchema),
});
export type PortfolioMetrics = z.infer<typeof portfolioMetricsSchema>;

export const portfolioDetailSchema = z.object({
  portfolio: portfolioSchema,
  holdings: z.array(holdingSchema),
  metrics: portfolioMetricsSchema,
});
export type PortfolioDetail = z.infer<typeof portfolioDetailSchema>;
